tangelo/toolboxes/circuits/qsp.py

```diff
@@ -0,0 +1,272 @@
+    pg = pyqsp.poly.PolySineTX()
+    pcoefs, _ = pg.generate(tau=tau,
+                            return_coef=True,
+                            ensure_bounded=True,
+                            return_scale=True, epsilon=eps)
+    for i in range(n_attempts):
+        try:
+            phiset2 = angle_sequence.QuantumSignalProcessingPhases(
+                pcoefs, eps=eps, suc=1-eps/10, method=method)
+        except (AngleFindingError, CompletionError):
+            if i == n_attempts-1:
+                raise RuntimeError("Imaginary phases calculation failed, increase n_attempts or eps")
+            else:
+                print(f"Attempt {i+2} for the imaginary phases")
+        else:
+            break
+    return phiset, phiset2
+
+
+def ham_sim_phases_QSPPACK(folder: str, tau: float, eps: float = 1.e-7) -> Tuple[List[float], List[float]]:
+    """Generate the phases required for QSP time-evolution using the QSPPACK package.
+
+    QSPPACK is an Matlab/Octave based package that calculates the phase factors. To run this function, a user needs to have:
+    1) Octave installed and accessible in path. Octave can be found at https://octave.org/
+    2) oct2py installed. pip install oct2py
+    3) QSPPACK downloaded in an accessible folder. Can be found at https://github.com/qsppack/QSPPACK/tree/dev
+
+    Args:
+        folder (str): The folder location of QSPPACK.
+        tau (float): The evolution time.
+        eps (float): The precision to calculate the factors.
+
+    Returns:
+        List[float]: The phases for Cos(Ht).
+        List[float]: The phases for i*Sin(Ht).
+    """
+    from oct2py import octave
+    from scipy.special import jn
+    octave.addpath(folder)
+    opts = octave.struct("criteria", eps)
+    maxorder = math.ceil(1.4*tau+np.log(1e14))
+    if maxorder % 2 == 1:
+        maxorder -= 1
+
+    coef = np.zeros((maxorder//2 + 1, 1), dtype=float, order='F')
+    for i in range(1, len(coef)+1):
+        coef[i-1][0] = (-1)**(i-1)*jn(2*(i-1), tau)
+
+    coef[0] = coef[0]/2
+    phi1, _ = octave.QSP_solver(coef, 0, opts, nout=2)
+
+    coef = np.zeros((maxorder//2+1, 1), dtype=float, order='F')
+    for i in range(1, len(coef)+1):
+        coef[i-1][0] = (-1)**(i-1)*jn(2*i-1, tau)
+    phi2, _ = octave.QSP_solver(coef, 1, opts, nout=2)
+    return list(phi1.flatten()), list(phi2.flatten())
+
+
+def zero_controlled_cnot(qubit_list: List[int], target: int, control: Union[int, List[int]]) -> Circuit:
+    """Return the circuit for a zero-controlled CNOT gate with possible extra controls on 1.
+
+    Args:
+        qubit_list (List[int]): The qubits controlled by zero.
+        target (int): The target qubit.
+        control (List[int] or int): The qubits controlled by one.
+
+    Returns:
+        Circuit: The zero controlled cnot circuit with possible extra controls on 1.
+    """
+    x_ladder = [Gate("X", q) for q in qubit_list]
+    gates = x_ladder + [Gate("CX", target=target, control=qubit_list+control)] + x_ladder
```

Comment:
Control can be an int or a union of int. What if control is an int here? Does this work ?